feat(Letterboxd): add presence

Skaytacium opened the pull request:
**Describe the changes in this pull request:**
Resolves #5230

<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

![image](https://user-images.githubusercontent.com/58383260/152532664-545fb83c-3112-4ca9-a6f5-ac0aee2893b8.png)
![image](https://user-images.githubusercontent.com/58383260/152532778-63e4ae54-c8fb-4f60-bb5b-36d56a8f1657.png)
![image](https://user-images.githubusercontent.com/58383260/152532853-4a1ad40d-0930-4bd2-8d23-ea0990c37f3f.png)
![image](https://user-images.githubusercontent.com/58383260/152532883-2810a5e6-8fc8-42b5-a3b2-fbaf79626d3b.png)
![image](https://user-images.githubusercontent.com/58383260/152532909-b08a9e9f-7799-4c43-a87b-c2e221dfe664.png)
![image](https://user-images.githubusercontent.com/58383260/152532938-791ccfc5-7525-47f8-83f0-109d85537dce.png)
![image](https://user-images.githubusercontent.com/58383260/152532961-b90e3d4b-6bd4-4499-8309-0fbad0751e85.png)
![image](https://user-images.githubusercontent.com/58383260/152534254-08a01bbc-f84e-40d6-905e-d16ddfbf3df6.png)
![image](https://user-images.githubusercontent.com/58383260/152532990-112800f4-dbf5-41fd-8067-bec191cedd60.png)
![image](https://user-images.githubusercontent.com/58383260/152533017-5b266571-6ca9-48a1-af33-18edc4b579da.png)
![image](https://user-images.githubusercontent.com/58383260/152535011-8c0cbbb9-2a1f-4610-a4bd-2b86393afde8.png)
![image](https://user-images.githubusercontent.com/58383260/152535032-b217ecbc-bbb8-4621-8d3c-5cc534f902a9.png)
![image](https://user-images.githubusercontent.com/58383260/152535194-c86a6470-fcd2-43c6-915a-51d8865ffdf6.png)
![image](https://user-images.githubusercontent.com/58383260/152535197-b833b43d-85d9-45ba-b04c-ea5dcc9ae8ef.png)
![image](https://user-images.githubusercontent.com/58383260/152535305-22a5f691-4017-4b47-ba01-f9a3369517f4.png)
![image](https://user-images.githubusercontent.com/58383260/152535311-2b93f40a-aff7-4c8a-8155-b961e359826c.png)
![image](https://user-images.githubusercontent.com/58383260/152535312-69711451-2a28-421a-9ec0-e7b5ed0412b9.png)


</details>

Commit:
feat(Letterboxd): add presence

Adds rich presence for Letterboxd as none is currently available (#5230)

Commit:
Merge branch 'main' of https://github.com/PreMiD/Presences

Commit:
feat(Letterboxd): improve presence

Simpler and cleaner rewrite with new and understandable
structure that allows easy integration of any new features.
Almost 100% site coverage added and added support for paths
that are not supported (shows basic info now).

Skaytacium committed:
Merge branch 'PreMiD:main' into main

Commit:
Merge branch 'main' of https://github.com/PreMiD/Presences

Commit:
fix(Letterboxd): clean up case

Cleans up a specific case (director/actor)

Commit:
Merge branch 'main' of https://github.com/Skaytacium/Presences

Commit:
feat(Letterboxd): adds statistics

Adds coverage for statistics

Commit:
fix(Letterboxd): fixes double title issue

Commit:
fix(Letterboxd): fixes formatting

Follows the formatting rules set by ESLint

EncryptedDev commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,450 @@
+	const pd: PresenceData = {
+		largeImageKey: "final",
+		startTimestamp: Date.now(),
+		buttons: [{ label: "View", url: window.location.href }]
```
View what...

EncryptedDev commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,450 @@
+	const path = document.location.pathname.slice(1).split("/");
+	path.pop();
+
+	const pd: PresenceData = {
```
```suggestion
	const presenceData: PresenceData = {
```

EncryptedDev reviewed (changes requested):
Look at my previous comments

Skaytacium committed:
Merge branch 'PreMiD:main' into main

Commit:
Merge branch 'main' of https://github.com/PreMiD/Presences

Commit:
feat(Letterboxd): adds button context

Adds auto-generated context for buttons when at specific links. Also
changes a variable name.

Commit:
Merge branch 'main' of https://github.com/Skaytacium/Presences

Skaytacium commented:
Added requested changes

<details>
<summary> Screenshots </summary>
<br>

![image](https://user-images.githubusercontent.com/58383260/152631663-adf10585-ce7a-41d6-9bb5-2f448139b578.png)
![image](https://user-images.githubusercontent.com/58383260/152631664-c6be93d5-54fa-4140-b033-a46db7b5759d.png)
![image](https://user-images.githubusercontent.com/58383260/152631665-9afbcfd2-1639-4278-be72-ad4fdd5cae53.png)
![image](https://user-images.githubusercontent.com/58383260/152631666-6211d5d7-1dac-4142-895f-bc39a042fa1c.png)
![image](https://user-images.githubusercontent.com/58383260/152631713-75d18e95-eaff-4ec7-a6e8-7e6d91beabef.png)
![image](https://user-images.githubusercontent.com/58383260/152631716-c30ac8fe-8cc9-46ac-8b07-16ac98bf4913.png)


</details>

Commit:
feat(Letterboxd): adds trailer support

Adds support for trailer watching

Skaytacium commented:
Are any further changes required?

Slowlife01 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,543 @@
+		// eslint-disable-next-line no-undefined
+		presenceData.buttons = undefined;
+
+	if (details !== presenceData.details) {
```
Pretty unnecessary

Skaytacium replied:
Its there for not messing up the `Date.now()` of "Time elapsed"

Skaytacium committed:
Update websites/L/Letterboxd/presence.ts

Co-authored-by: Slowlife <slowlife1165@gmail.com>

Commit:
Merge branch 'main' of https://github.com/PreMiD/Presences

Commit:
Merge branch 'main' of https://github.com/Skaytacium/Presences

Commit:
fix(Letterboxd): cleans up metadata

Removes unnecessary lines and makes the logo 512x512

Slowlife01 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,541 @@
+									}`;
+									presenceData.largeImageKey = getImageURLByAlt(title);
+									presenceData.smallImageKey = "final";
+									// eslint-disable-next-line no-undefined
+									presenceData.startTimestamp = undefined;
```
```suggestion
									delete presenceData.startTimestamp;
```

Slowlife01 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,541 @@
+function getImageURLByAlt(alt: string) {
+	const imgs = document.getElementsByTagName("img");
+
+	for (let i = 0; i < imgs.length; i++)
+		if (imgs[i].alt === alt) return imgs[i].src;
```
```js
Array.from(document.getElementsByTagName("img")).find(img => img.alt === alt)?.src;
```

Slowlife01 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,541 @@
+	itr: HTMLCollectionOf<T>,
+	fnc: (val: T, ind?: number) => boolean
+) {
+	for (let i = 0; i < itr.length; i++) if (fnc(itr[i], i)) return itr[i];
```
```js
Array.from(itr).find((element, ind) => fnc(element, ind));
```

Commit:
fix(Letterboxd): cleans up more code

More declarative functions, removes use of undefined.

EncryptedDev reviewed (changes requested):
Format the metadata to use tabs with Prettier. `prettier --write .`

Commit:
fix(Letterboxd): formats metadata

Bas950 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,539 @@
+const presence = new Presence({
+		clientId: "938732156346314795"
+	}),
+	user = document.cookie
+		.split(";")
+		.find(val => val.startsWith("letterboxd"))
+		.split("=")[1];
```
```suggestion
	user = document.cookie
		.split(";")
		.find(val => val.startsWith("letterboxd"))
		.split("=")[1],
	browsingTimestamp = Math.floor(Date.now() / 1000);
```

Bas950 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,539 @@
+
+	const presenceData: PresenceData = {
+		largeImageKey: "final",
+		startTimestamp: Date.now()
```
```suggestion
		startTimestamp: browsingTimestamp
```

Bas950 commented on websites/L/Letterboxd/presence.ts:
```diff
@@ -0,0 +1,539 @@
+
+	if (!(await presence.getSetting("show_buttons"))) delete presenceData.buttons;
+
+	if (details !== presenceData.details) {
```
now this can be removed

Skaytacium committed:
fix(Letterboxd): cleans timestamp 1

Co-authored-by: Bas950 <bas.van.zanten007@gmail.com>

Skaytacium committed:
fix(Letterboxd): cleans timestamp 2

Co-authored-by: Bas950 <bas.van.zanten007@gmail.com>

Commit:
fix(Letterboxd): cleans timestamp 3

Cleans the if statement at the end of the presence.ts by writing
timestamp to a variable

Commit:
Merge branch 'main' of https://github.com/Skaytacium/Presences

EncryptedDev commented:
Thanks for the PR!

elijaholmos commented:
awesome stuff, thank you!

xxKR15Hxx commented:
Hey, thanks for the PR really needed to show my friends want I'm doing on letterboxd and you fixed that,
Now all i need is friends
